src/AppInstallerCLICore/Workflows/PinFlow.cpp

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#include "pch.h"
+#include "Resources.h"
+#include "PinFlow.h"
+#include "TableOutput.h"
+#include "Microsoft/PinningIndex.h"
+#include "Microsoft/SQLiteStorageBase.h"
+#include "winget/RepositorySearch.h"
+
+using namespace AppInstaller::Repository;
+using namespace AppInstaller::Repository::Microsoft;
+
+namespace AppInstaller::CLI::Workflow
+{
+    namespace
+    {
+        // Creates a Pin appropriate for the context based on the arguments provided
+        Pinning::Pin CreatePin(Execution::Context& context, std::string_view packageId, std::string_view sourceId)
+        {
+            if (context.Args.Contains(Execution::Args::Type::BlockingPin))
+            {
+                return Pinning::Pin::CreateBlockingPin({ packageId, sourceId });

```

Comment:
Is there a reason we wouldn't want to record the currently installed version for a blocking pin or a pinning pin? Having the version would allow for heuristics on if pins are violated from installs outside of winget, and in the case of pinning pins could be updated when the user runs an upgrade with the proper argument. 

Especially as users may have side by side installs for a few packages, they may also want to have one of the versions pinned and another unpinned. It may be easier to record the version for now, even if nothing is done with it, to enable future functionality

Reply:
There was no specific reason for me to not include it, it was just that I didn't need it at the moment.

I see that in the spec the sample tables for listing pinning and blocking pins don't show versions, @yao-msft do you remember if there was a specific reason for that?

I think if we needed the version, it would be a gating pin, wouldn't it? The other two aren't saying "keep it in this version", but rather "don't touch them". What future functionality are you picturing that would benefit from having the version? The only I can think of is showing some message when we detect a version change outside of winget, but I don't know what we would like to do with that.

Reply:
> I think if we needed the version, it would be a gating pin, wouldn't it?

The way I see it, a pinning pin is a gating pin that is set to a specific version instead of a version range, which would exclude it from `upgrade --all`, and `upgrade <package>` would update the package and the version of the pin. Which would make a blocking pin a gating pin that is set to a specific version instead of a version range, and is blocked from upgrade entirely. 

> What future functionality are you picturing that would benefit from having the version?

You got the main one that I am thinking of, which would be to show a message when there is a version change outside of winget.

Another one that I see would be around import/export - Presume for a moment that a user has `Example.Package` with a blocking pin. On their computer, they have version 1.0.0 installed. The user then exports their package list, which should include the information that there is a blocking pin on the package.  They go to a new computer, and manually install `Example.Package` version 1.2.0. When running the package import, the package install would be skipped due to it already existing, but the blocking pin would still be added (unless I'm mistaken here and the pin wouldn't be added). Having the version associated with the blocking pin would allow for clearer messaging during import as to why a package was skipped, or why a pin wasn't added.

Another potential use case - I have `Example.Package` installed on my machine, I have version 1.0.0 installed and I have version 1.1.0 installed. For some workflows, I specifically need version 1.0.0, but for other workflows I need the latest version. It would be beneficial to be able to add a blocking pin on version 1.0.0 to ensure that it is never upgraded, while still allowing version 1.1.0 to be upgraded to 1.1.1 or 1.2.0. 

Another use case could be a future pinning type - Locking. This future type might prevent both upgrade and uninstall. 

Directly mentioned in the spec is a future consideration for integration with the Dependencies flow. It could make it easier for dependencies to check the version of the blocking pin, rather than having to check the installed version if there is a blocking pin.

Reply:
Updated to keep the version in all cases. For now I'm just keeping it as a string since I'm not using it yet.

Reply:
> That's also the reason "pinning from installed source" is put in the future consideration section since this depends on underlying Composite package improvement.
> 
I see. It wasn't clear what exactly 'installed source' meant, I think I understand now

> In the spec time, I was thinking rather than recording something unknown or uncertain, we would leave it until we have better underlying framework code support.
>
If multiple results are returned for an installed version, the field could be left empty; If a single version is installed, then it could be populated? Unless I'm misunderstanding how the composite packages work

Reply:
> If multiple results are returned for an installed version, the field could be left empty; If a single version is installed, then it could be populated? Unless I'm misunderstanding how the composite packages work

That's exactly the issue we have with current CompositePackage implementation. The CompositePackage returned by the search result always return at most 1 installed version, and we don't know if there's only 1 installed version vs. there are multiple installed versions. And I think we'd better deal this after the underlying CompositePackage is improved.

Reply:
Changed to remove the version from non-gating pins